### PR TITLE
Add markdownlint support

### LINKED ISSUE #453: Standarize Markdownlinter
It would be desirable to lint markdown for the documentation and have it in a standard format.

### Specification

- Update all files to be lint to markdown
- Implement the [markdown lint cli](https://github.com/igorshubovych/markdownlint-cli) in the devops pipeline

### PR DESCRIPTION
**Pull request by adamconnelly**
Fixes #453

## Proposed Changes

- Added markdownlint via npm, and added a basic configuration file.
  At the moment I've just setup the defaults, but disabled the line
  length rule for code blocks and tables.
- Updated the contributing guide to include information on running it.
- Added a new stage to the build pipeline to run the linting.

@tomkerkhove I didn't want to take this too far, but I've got another couple of suggestions:

- We can use [Husky](https://github.com/typicode/husky) to create shared git hooks. What this would let us do is define a pre-commit hook that automatically runs markdownlint. This means that people get notified immediately if they've made a linting error, without having to wait for a build to run on GitHub.
- We could potentially use [Prettier](https://prettier.io/blog/2017/11/07/1.8.0.html) to automatically format our markdown. What this would mean is that if you made a simple mistake that it can fix like a line being longer than 80 characters, we'd run Prettier as part of the Husky pre-commit hook, and automatically fix any simple problems before linting. This would mean less manual fixing of simple formatting issues.

### CODE REVIEW

**Comment by adamconnelly** [comment]
Sorry about all the noise in this PR - it took me a while to get the command running properly in Azure Pipelines.

Here's an example failing build: https://tomkerkhove.visualstudio.com/Promitor/_build/results?buildId=1511&view=logs&j=a0162c90-9618-5d58-2cc8-c63b24381f83&t=50aea722-6967-5a54-fc98-844cbba5a109&l=15

File: build/azure-devops/scraper-ci.yml [hunk 1]
```diff
@@ -77,6 +77,20 @@ stages:
...
+  jobs:
+   - job: RunLinters
+     displayName: Run Linters
+     condition: succeeded()
+     pool:
+       vmImage: ubuntu-16.04
+     steps:
+     - script: npm install
+       displayName: 'Install npm Packages'
+     - script: npm run --silent markdownlint
```
Comment (tomkerkhove) [inline build/azure-devops/scraper-ci.yml]: What does the `--silent` do?

File: build/azure-devops/scraper-ci.yml [hunk 2]
```diff
@@ -77,6 +77,20 @@ stages:
          solutionOrProjectPath: 'src/Promitor.sln'
          failBuildLevelSelector: 'Warning'
          commandLineInterfacePath: '$(Build.SourcesDirectory)/Lib/Resharper'
+- stage: Linting
+  displayName: Linting
```
Comment (tomkerkhove) [inline build/azure-devops/scraper-ci.yml]: Let's call this stage `Docs` maybe?

File: build/azure-devops/scraper-ci.yml [hunk 3]
```diff
@@ -77,6 +77,20 @@ stages:
...
+  dependsOn: []
+  jobs:
+   - job: RunLinters
+     displayName: Run Linters
```
Comment (tomkerkhove) [inline build/azure-devops/scraper-ci.yml]: Given we only have this linter it's maybe better to rename this?
```suggestion
     displayName: Run Markdown Linter
```

File: charts/promitor-agent-scraper/README.md [hunk 4]
```diff
@@ -1,17 +1,21 @@
...
 
-[Promitor](https://promitor.io/) is an Azure Monitor scraper for Prometheus providing a scraping endpoint for Prometheus that provides a configured subset of Azure Monitor metrics.
+[Promitor](https://promitor.io/) is an Azure Monitor scraper for Prometheus providing
+a scraping endpoint for Prometheus that provides a configured subset of Azure
+Monitor metrics.
 
 ## TL;DR
 
 ```console
-$ helm repo add promitor https://promitor.azurecr.io/helm/v1/repo
```
Comment (tomkerkhove) [inline charts/promitor-agent-scraper/README.md]: Is this because it's blocked or a suggestion?

**Comment by tomkerkhove** [comment]
Thanks for the PR @adamconnelly! Is there a reason why the guideline is to only have x characters on one line?

> * We can use [Husky](https://github.com/typicode/husky) to create shared git hooks. What this would let us do is define a pre-commit hook that automatically runs markdownlint. This means that people get notified immediately if they've made a linting error, without having to wait for a build to run on GitHub.
> * We could potentially use [Prettier](https://prettier.io/blog/2017/11/07/1.8.0.html) to automatically format our markdown. What this would mean is that if you made a simple mistake that it can fix like a line being longer than 80 characters, we'd run Prettier as part of the Husky pre-commit hook, and automatically fix any simple problems before linting. This would mean less manual fixing of simple formatting issues.

I'd suggest to create a dedicated issue for this discussion, is that ok for you?

But for now locally this will not be enforced until you create a PR where the linting is happening, is that correct? If so, do we still need the edit to `.github/CONTRIBUTING.md`?

**Comment by tomkerkhove** [comment]
> Sorry about all the noise in this PR - it took me a while to get the command running properly in Azure Pipelines.
> 
> Here's an example failing build: [tomkerkhove.visualstudio.com/Promitor/_build/results?buildId=1511&view=logs&j=a0162c90-9618-5d58-2cc8-c63b24381f83&t=50aea722-6967-5a54-fc98-844cbba5a109&l=15](https://tomkerkhove.visualstudio.com/Promitor/_build/results?buildId=1511&view=logs&j=a0162c90-9618-5d58-2cc8-c63b24381f83&t=50aea722-6967-5a54-fc98-844cbba5a109&l=15)

No worries, thanks!

File: build/azure-devops/scraper-ci.yml (hunk 1 above)
Comment (adamconnelly) [reply build/azure-devops/scraper-ci.yml]: It's just to get rid of the noise that npm outputs when a command fails. It's easier to explain with an example, so here's what it's like without `--silent`:

```shell
 $ npm run markdownlint                                                       /c/Users/AdamConnelly/github.com/adamrpconnelly/promitor

> promitor@1.0.0 markdownlint C:\Users\AdamConnelly\github.com\adamrpconnelly\promitor
> markdownlint -c .markdownlint.json --ignore node_modules .

README.md: 28: MD013/line-length Line length [Expected: 80; Actual: 142]
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! promitor@1.0.0 markdownlint: `markdownlint -c .markdownlint.json --ignore node_modules .`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the promitor@1.0.0 markdownlint script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\AdamConnelly\AppData\Roaming\npm-cache\_logs\2019-10-12T16_05_15_723Z-debug.log
```

But if we include the `--silent` flag, we just get the actual errors output, and we don't get all the `npm ERR!` lines that don't really add any value:

```shell
$ npm run --silent markdownlint                                            /c/Users/AdamConnelly/github.com/adamrpconnelly/promitor 1
README.md: 28: MD013/line-length Line length [Expected: 80; Actual: 142]
```

File: build/azure-devops/scraper-ci.yml (hunk 2 above)
Comment (adamconnelly) [reply build/azure-devops/scraper-ci.yml]: Sure - I was getting ahead of myself 😄

File: build/azure-devops/scraper-ci.yml (hunk 3 above)
Comment (adamconnelly) [reply build/azure-devops/scraper-ci.yml]: No problem.

File: charts/promitor-agent-scraper/README.md (hunk 4 above)
Comment (adamconnelly) [reply charts/promitor-agent-scraper/README.md]: It's because the command has no output. The default rules have an error if you put a dollar in front of commands with no output.

For example, the following is ok:

```console
helm repo add promitor https://promitor.azurecr.io/helm/v1/repo
```

So is the following:

```console
$ helm repo add promitor https://promitor.azurecr.io/helm/v1/repo
"promitor" has been added to your repositories
```

But the following causes an error:

```shell
$ helm repo add promitor https://promitor.azurecr.io/helm/v1/repo
```

I actually think it makes sense, but I'm happy to disable that rule if you want. I just figured I'd start with the defaults.

**Comment by adamconnelly** [comment]
I think the 80 character limit is just to make the files readable when you're viewing them in a text editor, rather than in a browser. If you view the docs in GitHub or on the site, it word wraps and enforces a gutter on either side to make them readable, but if you view them in a text editor it won't necessarily do that, so you can end up with really long lines.

But it's just a default setting, so we can change it if you want.

File: charts/promitor-agent-scraper/README.md (hunk 4 above)
Comment (tomkerkhove) [reply charts/promitor-agent-scraper/README.md]: No worries, I was just curious - Thanks for explaining!

**Comment by adamconnelly** [comment]
@tomkerkhove I've just pushed another commit that reverts the change to CONTRIBUTING.md, renamed the pipeline steps and adds information on running markdownlint to the docs README.

You're correct - it'll only run in CI, or if you manually run `npm run markdownlint` for now.

File: build/azure-devops/scraper-ci.yml (hunk 1 above)
Comment (tomkerkhove) [reply build/azure-devops/scraper-ci.yml]: Sounds good, thanks!

File: README.md [hunk 5]
```diff
@@ -33,15 +46,25 @@ Are you having issues or feature requests? Feel free to [let us know](https://gi
...
 
 ## Get in touch
-Do you have a security issue to report or just want to privately contact me? Feel free to [write me an email](mailto:kerkhove.tom@gmail.com)
 
-## Thank you!
-We'd like to thank all the services, tooling & NuGet packages that support us - [Thank you](https://promitor.io/thank-you)!
+Do you have a security issue to report or just want to privately contact me? Feel
+free to [write me an email](mailto:kerkhove.tom@gmail.com)
+
+## Thank you
```
Comment (tomkerkhove) [inline README.md]: Was the `!` on purpose removed?

Comment (adamconnelly) [reply README.md]: Yeah - it's disallowed by https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md#md026---trailing-punctuation-in-header. Unfortunately they don't really provide much of a reason for why certain rules are enforced.

I can just disable that rule and add the `!` back in if you want, or we could disable it for this one item. What's your thoughts?

Comment (tomkerkhove) [reply README.md]: Nah that's ok, thanks for pointing to that repo!

**Comment by tomkerkhove** [comment]
Thanks Adam!
